lib/spack/spack/spec.py

```diff
@@ -1296,22 +1297,35 @@ def prefix(self):
     def prefix(self, value):
         self._prefix = Prefix(value)
 
-    def dag_hash(self, length=None):
+    def dag_hash(self, length=None, all_deps=False):
         """Return a hash of the entire spec DAG, including connectivity."""
-        if self._hash:
-            return self._hash[:length]
-        else:
-            yaml_text = syaml.dump(
-                self.to_node_dict(), default_flow_style=True, width=maxint)
-            sha = hashlib.sha1(yaml_text.encode('utf-8'))
+        if not self.concrete:
+            h = self._dag_hash(all_deps=all_deps)
+            # An upper bound of None is equivalent to len(h). An upper bound of
+            # 0 produces the empty string
+            return h[:length]
 
-            b32_hash = base64.b32encode(sha.digest()).lower()
-            if sys.version_info[0] >= 3:
-                b32_hash = b32_hash.decode('utf-8')
+        if not self._hash:
+            self._hash = self._dag_hash(all_deps=False)
 
-            if self.concrete:
-                self._hash = b32_hash
-            return b32_hash[:length]
+        if not self._build_hash:
+            self._build_hash = self._dag_hash(all_deps=True)
```

Comment:
Just use the `full_hash()` method here and memoize the full hash separately.  *or* handle the memoization with a dict of hashes by deptype, e.g.:

```python
deptypes = ('build',  'link')
self._hashes[deptypes] = self._dag_hash(deptype=deptypes)
```